llvm/lib/Analysis/CFG.cpp

```diff
@@ -204,6 +204,91 @@ bool llvm::isPotentiallyReachableFromMany(
   return false;
 }
 
+bool llvm::isManyPotentiallyReachableFromMany(
+    SmallVectorImpl<BasicBlock *> &Worklist,
+    const SmallPtrSetImpl<const BasicBlock *> &StopSet,
+    const SmallPtrSetImpl<BasicBlock *> *ExclusionSet, const DominatorTree *DT,
+    const LoopInfo *LI) {
+  // When a stop block is unreachable, it's dominated from everywhere,
+  // regardless of whether there's a path between the two blocks.
+  SmallPtrSet<const BasicBlock *, 32> StopBBReachable;
+  for (auto *BB : StopSet) {
+    if (DT && DT->isReachableFromEntry(BB))
+      StopBBReachable.insert(BB);
+  }
+
+  // We can't skip directly from a block that dominates the stop block if the
+  // exclusion block is potentially in between.
+  if (ExclusionSet && !ExclusionSet->empty())
+    DT = nullptr;
+
+  // Normally any block in a loop is reachable from any other block in a loop,
+  // however excluded blocks might partition the body of a loop to make that
+  // untrue.
+  SmallPtrSet<const Loop *, 8> LoopsWithHoles;
+  if (LI && ExclusionSet) {
+    for (auto *BB : *ExclusionSet) {
+      if (const Loop *L = getOutermostLoop(LI, BB))
+        LoopsWithHoles.insert(L);
+    }
+  }
+
+  llvm::DenseMap<const BasicBlock *, const Loop *> StopLoops;
+  for (auto *StopBB : StopSet)
+    StopLoops[StopBB] = LI ? getOutermostLoop(LI, StopBB) : nullptr;
```

Comment:
Why do we need this to be a map? Shouldn't this also be a set of StopLoops and then just check contains below?

Reply:
We need the corresponding `Loop *` for each `StopBB`.

Calling `getOutermostLoop(...)` isn't cheaper - we perform a map lookup and repeatedly call `Loop::getParent(...)` until we get the outermost loop.

Reply:
Fixed.

Reply:
I don't get it -- why can't we do something like this? https://github.com/llvm/llvm-project/commit/1c48a3a138c4e0126abf4cc255c56fd5b65129d5

Reply:
Sorry, I misread your comment. Yeah, that would work. Thanks for the suggestion!